Add transaction deny support

lxfind opened the pull request:
This PR adds ability to deny a transaction at signing stage. The deny policy can be controlled through the node's config file.
TODO: Add comments and tests

sblackshear commented on crates/sui-core/src/transaction_signing_filter.rs:
```diff
@@ -0,0 +1,165 @@
+    }
+    deny_if_true!(
+        deny_map.contains(&tx_data.sender()),
+        format!("Access to account address {:?} is denied", tx_data.sender())
```
```suggestion
        format!("Access to account address {:?} is temporarily denied", tx_data.sender())
```

sblackshear commented on crates/sui-core/src/transaction_signing_filter.rs:
```diff
@@ -0,0 +1,165 @@
+    deny_if_true!(
+        deny_map.contains(&tx_data.gas_owner()),
+        format!(
+            "Access to account address {:?} is denied",
```
```suggestion
            "Access to account address {:?} is temporarily denied",
```

sblackshear commented on crates/sui-core/src/transaction_signing_filter.rs:
```diff
@@ -0,0 +1,165 @@
+        let id = object_kind.object_id();
+        deny_if_true!(
+            deny_map.contains(&id),
+            format!("Access to input object {:?} is not allowed", id)
```
```suggestion
            format!("Access to input object {:?} is temporarily not allowed", id)
```

sblackshear commented on crates/sui-core/src/transaction_signing_filter.rs:
```diff
@@ -0,0 +1,165 @@
+        );
+        deny_if_true!(
+            filter_config.shared_object_disabled() && object_kind.is_shared_object(),
+            "Usage of shared object in transactions is not allowed"
```
```suggestion
            "Usage of shared object in transactions is temporarily not allowed"
```

sblackshear commented on crates/sui-core/src/transaction_signing_filter.rs:
```diff
@@ -0,0 +1,165 @@
+        deny_if_true!(
+            deny_map.contains(&dep),
+            format!(
+                "Package {:?} cannot be accessed either directly or indirectly",
```
```suggestion
                "Access to package {:?} is temporarily disabled--it cannot be accessed either directly or indirectly",
```

lxfind commented:
All tests added

amnn commented on crates/sui-config/src/transaction_deny_config.rs:
```diff
@@ -0,0 +1,138 @@
+pub struct TransactionDenyConfig {
+    /// A list of object IDs that are not allowed to be accessed/used in transactions.
+    /// Note that since this is checked during transaction signing, only root object ids
+    /// are supported here (i.e. no child-objects).
```
Also no wrapped objects, and practically speaking any object whose type has `store` can also skirt around this restriction (not an issue, but worth documenting).

amnn commented on crates/sui-core/src/transaction_signing_filter.rs:
```diff
@@ -0,0 +1,171 @@
+                            },
+                        })?;
+                // linkage_table contains all transitive dependencies of the package.
+                dependencies.extend(package.linkage_table().keys().copied());
```
You don't want the `keys` here, you want the `upgrade_id` in the values -- this is the ID of the package in storage that actually gets loaded during execution (and corresponds to the value that comes from `deps` in the `Publish` and `Upgrade` transactions).

Having said that, maybe you want to be able to prevent calls to all versions of a package?  In which case you should use the `keys()` of the linkage table and the `original_package_id()` of the package being called as well.

amnn commented on crates/sui-core/src/transaction_signing_filter.rs:
```diff
@@ -0,0 +1,171 @@
+    if !filter_config.package_publish_disabled() && !filter_config.package_upgrade_disabled() {
+        return Ok(());
+    }
+    for command in tx_data.kind().iter_commands() {
```
Does it make sense to add a facility to disable each command? (e.g. also `Split`, `Merge` and `Transfer`?). It's quite an extreme move, and I can't think of an immediate reason to use it, but I'd hate to need it and not have it.

amnn commented on crates/sui-core/src/unit_tests/transaction_deny_tests.rs:
```diff
@@ -0,0 +1,238 @@
+}
+
+#[tokio::test]
+async fn test_package_denied() {
```
Probably worth adding a test where you deny access to an upgraded package but not the original version?  That's the path that would be affected by the linkage table change I mentioned earlier.

amnn reviewed (commented):
There's an issue with how upgraded packages are denied at the moment, but otherwise this looks good -- I think you could go two ways with that filter (deny specific packages, or deny families of packages), with pro's and con's on each side.

lxfind commented:
Addressed the feedback by only comparing against the upgraded version.
Still need to add tests for this, which will take a little while since it's non-trivial to setup.

amnn commented on crates/sui-core/src/transaction_signing_filter.rs:
```diff
@@ -0,0 +1,172 @@
+    Ok(())
+}
+
+fn check_singers(filter_config: &TransactionDenyConfig, tx_data: &TransactionData) -> SuiResult {
```
nit, spelling
```suggestion
fn check_signers(filter_config: &TransactionDenyConfig, tx_data: &TransactionData) -> SuiResult {
```

amnn reviewed (approved):
LGTM, thanks @lxfind !
